Add RFC for creation and use of NUMA-constrained arenas (#1559)
uxlfoundation/oneTBB: rfcs/proposed/numa_support/numa-arenas-creation-and-use.org

same number of ~tbb::task_group~ objects, with which the oneTBB tasks are going
to be associated. The ~tbb::task_group~ objects are needed because they allow
waiting for the work completion as the ~tbb::task_arena~ class does not provide
synchronization semantics on its own. Then the work gets submitted in each of
arena objects, and waited upon their finish at the end.

** Interface issues and inconveniences:
- [0] - Getting the number of NUMA nodes is not the task by itself, but rather a
  necessity to know how many objects to initialize further.
- [1] - Explicit step for creating the number of ~tbb::task_arena~ objects per
  each NUMA node. Note that by default the arena objects are constructed with a
  slot reserved for master thread, which in this particular example usually
  results in undersubscription issue as the master thread can join only one
  arena at a time to help with work processing.
- [2] - Separate step for instantiation the same number of ~tbb::task_group~
  objects, in which the actual work is going to be submitted. Note that user
  also needs to make sure the size of ~arenas~ matches the size of

>   ~task_groups~.

Reviewer: I actually don't mind the repetition. Its easy to read over the "same number" without recognizing there's a potential error if the sizes of these two vectors don't match. So, in my opinion, the repetition highlights the potential danger here.

> - [2] - Separate step for instantiation the same number of ~tbb::task_group~
>   objects, in which the actual work is going to be submitted. Note that user
>   also needs to make sure the size of ~arenas~ matches the size of
>   ~task_groups~.

Reviewer 2: Fair enough, then I would rephrase as:
```suggestion
- [2] - The necessity to instantiate the same number of ~tbb::task_group~
  objects for the actual work to be submitted; that is, the size of ~task_groups~
  must match the size of ~arenas~.
```

- [3] - Actual tying of ~tbb::task_arena~ instances with corresponding NUMA
  nodes. Note that user needs to make sure the indices of ~tbb::task_arena~

>   objects match corresponding indices of NUMA nodes.

Reviewer 2: The point about the need to match the indices is kind of strange. A single loop that works with several arrays/vectors is a typical pattern, you just use the loop index consistently. Moreover, with modern C++ you can rewrite the loop to not have any indices at all, e.g.

```
std::vector<tbb::numa_node_id> numa_indexes = tbb::info::numa_nodes();
std::vector<tbb::task_arena> arenas;      // note that the size is not set
std::vector<tbb::task_group> task_groups; // same for task groups

for (auto idx: numa_indexes) {
    arenas.emplace_back( tbb::task_arena::constraints(idx) );
    task_groups.emplace_back();
    arenas.back().execute([&tg = task_groups.back()]{
        tg.run([]{/*some parallel stuff*/});
    });
}
```

If you meant something else, perhaps try explaining it better.

Reviewer: I think the point here might be not that there's a safe way to do it and a well written piece of code will do it the safe way. I think the point is that a not-so-well-written might mess it up. But this specific example (with everything in a loop body) doesn't provide much room for mismatched indices, so it doesn't seem at all likely in this case. In the more general case of task_arenas with a matching number of task_groups, it could be possible to mismatch them.

Reviewer 2: The sentence says "the user **needs to** make sure ...", and my point is - no, users do not necessarily need to. And we agree that this specific pattern "doesn't provide much room for mismatched indices".

My point is that we should not paint the usage worse than it really is.

- [4] - Actual work submission point. It is relatively easy to make a mistake
  here by using the ~tbb::task_arena::enqueue~ method instead. In this case not
  only the work submission might be done after the synchronization point [5],
  but also the loop counter ~j~ can be mistakenly captured by reference, which
  at least results in submission of the work into incorrect ~tbb::task_group~,
  and at most a segmentation fault, since the loop counter might not exist by

>   the time the functor starts its execution.

Reviewer: Since we're highlighting possible points of failure, I suppose we don't want to bring up the safer enqueue deferred tasks patten, right?

Reviewer 2: Well, I would really do want to bring it up.
